src/Uns.ts

```diff
@@ -255,26 +257,74 @@ export default class Uns extends NamingService {
     return !(await this.isRegistered(domain));
   }
 
-  async registryAddress(domain: string): Promise<string> {
-    if (!this.checkDomain(domain)) {
+  async registryAddress(domainOrNamehash: string): Promise<string> {
+    if (
+      !this.checkDomain(domainOrNamehash, domainOrNamehash.startsWith('0x'))
+    ) {
       throw new ResolutionError(ResolutionErrorCode.UnsupportedDomain, {
-        domain,
+        domain: domainOrNamehash,
       });
     }
-
-    const tld = domain.split('.').pop();
-    const [address] = await this.readerContract.call('registryOf', [
-      this.namehash(tld!),
-    ]);
-
+    const namehash = domainOrNamehash.startsWith('0x')
+      ? domainOrNamehash
+      : this.namehash(domainOrNamehash);
+    const [address] = await this.readerContract.call('registryOf', [namehash]);
     if (address === NullAddress) {
-      throw new ResolutionError(ResolutionErrorCode.UnsupportedDomain, {
-        domain,
+      throw new ResolutionError(ResolutionErrorCode.UnregisteredDomain, {
+        domain: domainOrNamehash,
       });
     }
     return address;
   }
 
+  async getDomainFromTokenId(tokenId: string): Promise<string> {
+    const registryAddress = await this.registryAddress(tokenId);
+    const registryContract = new EthereumContract(
+      registryAbi,
+      registryAddress,
+      this.provider,
+    );
+    const startingBlock = this.getStartingBlockFromRegistry(registryAddress);
+    const newURIEvents = await registryContract.fetchLogs(
+      'NewURI',
+      tokenId,
+      startingBlock,
+    );
+    if (!newURIEvents || newURIEvents.length === 0) {
+      throw new ResolutionError(ResolutionErrorCode.UnregisteredDomain, {
+        domain: `with tokenId ${tokenId}`,
+      });
+    }
+    const rawData = newURIEvents[newURIEvents.length - 1].data;
+    const decoded = Interface.getAbiCoder().decode(['string'], rawData);
+    return decoded[decoded.length - 1];
+  }
+
+  private getStartingBlockFromRegistry(registryAddress: string): string {
+    const contractDetails = Object.values(UnsConfig?.networks).reduce(
```

Comment:
Please check that this method works properly in case of custom network. Add tests for this.